### PR TITLE
Fixes #20580: module node_settings query option does not work

### PR DESCRIPTION
**Pull request by Fdall**
https://issues.rudder.io/issues/20580

### CODE REVIEW

**Comment by Fdall** [comment]
PR updated with a new commit

File: ci/ansible-test.Dockerfile [hunk 1]
```diff
@@ -4,7 +4,13 @@ COPY ci/user.sh .
 RUN ./user.sh $USER_ID
 
 RUN apt-get -y update && \
-    apt-get install -y ansible git python3-pip shellcheck
+    apt-get install -y ansible git python3-pip
 RUN pip install pycodestyle voluptuous yamllint
 
+RUN mkdir -p /tmp/ansible_collections/rudder/rudder
```
Comment (Fdall) [inline ci/ansible-test.Dockerfile:10]: This shoudl not be coommited, it is a local change

**Comment by Fdall** [comment]
PR updated with a new commit

**Comment by Fdall** [comment]
PR rebased

**Comment by Fdall** [comment]
PR updated with a new commit

File: plugins/modules/node_settings.py [hunk 2]
```diff
@@ -264,42 +296,21 @@ def _send_request(
...
-                    headers=self.headers,
-                    params=params,
-                    verify=verify,
+            resp = (
+                open_url(
+                    full_url,
+                    headers=headers,
+                    validate_certs=self.validate_certs,
+                    method=method,
+                    data=data,
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:305]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 3]
```diff
@@ -264,42 +296,21 @@ def _send_request(
...
+                .read()
+                .decode('utf8')
             )
-
-        status_code = resp.status_code
-        if status_code == 200:
-            return self._module.from_json(resp.content)
-        else:
+            return self._module.from_json(resp)
+        except Exception as e:
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:311]: Can you give a more explicit name than "e" for the error storage?

File: plugins/modules/node_settings.py [hunk 4]
```diff
@@ -338,7 +349,6 @@ def set_node_settings(self, node_id):
                 path='/api/latest/nodes/{node_id}'.format(node_id=node_id),
                 data=self.settings_to_set,
                 headers=self.headers,
-                serialize_json=True,
                 method='POST',
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:352]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 5]
```diff
@@ -350,44 +360,38 @@ def evaluate_node_query(self):
...
-            'composition': self.query_composition,
-        }
-
-        where_json_struct = {
-            'objectType': self.where_object_type,
-            'attribute': self.where_attribute,
-            'comparator': self.where_comparator,
-            'value': self.where_value,
+            'select': query['select'],
+            'composition': query['composition'],
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:367]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 6]
```diff
@@ -350,44 +360,38 @@ def evaluate_node_query(self):
...
-            path=url,
-            headers=self.headers,
-            serialize_json=False,
-            params=params,
+        nodes = self._send_request(
             method='GET',
-        )
+            path='/api/latest/nodes' + url_query,
+            data={},
+            headers=self.headers,
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:376]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 7]
```diff
@@ -350,44 +360,38 @@ def evaluate_node_query(self):
...
 
 def main():
     # Definition of the arguments and options
     # of the 'node_settings' module
+    where_object = dict(
+        object_type=dict(type='str', required=False),
+        attribute=dict(type='str', required=False),
+        comparator=dict(
+            type='str',
+            required=False,
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:391]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 8]
```diff
@@ -444,18 +448,10 @@ def main():
...
-                            attribute=dict(type='str', required=False),
-                            comparator=dict(
-                                type='str',
-                                choices=['and', 'or'],
-                                required=False,
-                            ),
-                            value=dict(type='str', required=False),
-                        ),
+                        elements='dict',
+                        options=where_object,
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:454]: You can delete the "," which serves no purpose here.

File: plugins/modules/node_settings.py [hunk 9]
```diff
@@ -466,30 +462,33 @@ def main():
...
+    failed = bool(errors)
     module.exit_json(
-        failed=False,
+        failed=failed,
         changed=changed,
         meta=module.params,
         audited_settings=rudder_node_iface.settings_to_set,
         nodes=impacted_nodes,
+        query=query,
+        errors=errors,
```
Comment (julienbriault) [inline plugins/modules/node_settings.py:491]: You can delete the "," which serves no purpose here.

File: plugins/modules/server_settings.py [hunk 10]
```diff
@@ -108,31 +108,23 @@ def _send_request(self, url, data=None, headers=None, method='GET'):
...
+        try:
+            resp = open_url(
+                full_url,
+                headers=headers,
+                validate_certs=self.validate_certs,
+                method=method,
+                data=data
+            ).read().decode('utf8')
+            return self._module.from_json(resp)
+        except Exception as e:
```
Comment (julienbriault) [inline plugins/modules/server_settings.py:123]: Can you give a more explicit name than "e" for the error storage?

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
+from __future__ import (absolute_import, division, print_function)
```
Comment (julienbriault) [inline tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:1]: Is the fact that there are no shebang on purpose?

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+class TestStringMethods(unittest.TestCase):
+    @parameterized.expand(
+        [
+            [
+                [
+                    {
+                        'object_type': 'node',
+                        'attribute': 'nodeHostname',
+                        'comparator': 'eq',
+                        'value': 'my_machine.my_domain',
```
Comment (julienbriault) [inline tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:17]: You can delete the "," which serves no purpose here.

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+                    }
+                ],
+                'where=[{"objectType":"node","attribute":"nodeHostname","comparator":"eq","value":"my_machine.my_domain"}]',
+            ],
+            [
+                {
+                    'object_type': 'node',
+                    'attribute': 'OS',
+                    'comparator': 'eq',
+                    'value': 'Linux',
```
Comment (julienbriault) [inline tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:27]: Same

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+                },
+                {
+                    'object_type': 'node',
+                    'attribute': 'osFullName',
+                    'comparator': 'regex',
+                    'value': '.*Linux.*',
```
Comment (julienbriault) [inline tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:33]: Same

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+                },
+                {
+                    'object_type': 'memoryPhysicalElement',
+                    'attribute': 'quantity',
+                    'comparator': 'gteq',
+                    'value': '1',
```
Comment (julienbriault) [inline tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:39]: Same

**Review by julienbriault — COMMENTED** [review]
I have added comments on things to correct/improve.

File: tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py (hunk 11 above)
Comment (Fdall) [reply tests/unit/plugins/modules/test_node_settings_should_format_well_the_queries.py:1]: Yes it is not meant to be executable, it should be run through pytest or ansible-test

**Comment by Fdall** [comment]
PR updated with a new commit

**Review by julienbriault — COMMENTED** [review]
Everything is ok!

**Pull request merged**